tests/utils/test_readers_safedir.py

```
@@ -0,0 +1,1443 @@
+        None for genuinely unknown formats). After PR3d, every reader the
+        codebase supports has a SafeDir mapping, so this exercises the
+        truly-unsupported branch with a synthetic extension.
+        """
+        (tmp_path / "data.unknownext").write_text("dummy")
+        with SafeDir.open_root(tmp_path) as sd:
+            assert read_file_via_safedir(sd, "data.unknownext") is None
+
+    def test_rejects_bad_name(self, tmp_path: Path) -> None:
+        """Component-name validation rides on SafeDir.open_for_reader."""
+        with SafeDir.open_root(tmp_path) as sd:
+            with pytest.raises(ValueError):
```

Ruff (0.15.13)

[warning] Use a single `with` statement with multiple contexts instead of nested `with` statements

Combine `with` statements

(SIM117)